Implement {to,from}{Node,Edge}Address

decentralion opened the pull request:
This commit implements Node/Edge addresses, and helper functions for
generating and manipulating them.

Test plan: Unit tests included.

Paired with @wchargin

decentralion commented:
BTW, when we paired this commit, we forgot to consider the case where `toNodeAddress` is passed an empty array. This resulted in a node address without any null separators, which `fromNodeAddress` interpreted as a  malformed node address. I've changed the behavior so `toNodeAddress` errors on an empty array input, and tested this.

wchargin commented:
Oh, I see your comment now:

> we forgot to consider the case where toNodeAddress is passed an empty
> array. This resulted in a node address without any null separators,
> which fromNodeAddress interpreted as a malformed node address.

That’s just an implementation error, then. We should fix the
implementation, not disallow the empty array.

I see two reasonable options:

  - We could change the representation from “NUL between components” to
    “NUL after each component”, so that `["foo"]` becomes `"N\0foo\0"`
    and `[]` becomes `"N\0"`.

  - We could keep the current representation and fix the `assert`
    functions, to check that the string’s first NUL-separated component
    is `N`: i.e., that the string either is literally `"N"` or starts
    with `"N\0"`. This should be equivalent to checking that
    `theString.split("\0")[0] === "N"`, which is the intended semantics
    anyway—the fact that this had the wrong behavior on the empty
    array is an error.

Either of these is fine with me.

decentralion commented:
Makes sense. I changed the implementation to always include a NUL separator after every path component; this involved re-writing the methods. I also refactored the test code. PTAL.

wchargin commented on src/v3/core/graph.test.js:
```diff
@@ -0,0 +1,94 @@
+        );
+      });
+      describe("compose to identity", () => {
+        function checkIdentity(name, example) {
```
+1

wchargin commented on src/v3/core/graph.js:
```diff
@@ -20,22 +21,72 @@ export type NeighborsOptions = {|
-  const _ = arr;
-  throw new Error("toNodeAddress");
+  assertAddressArray(arr);
+  if (arr.length > 0) {
```
Okay. This implementation isn’t quite as neat as I’d like (something
like `["N", ...arr, ""].join(SEPARATOR)` would be better, in my
opinion), but not a big deal.

wchargin commented on src/v3/core/graph.js:
```diff
@@ -20,22 +21,72 @@ export type NeighborsOptions = {|
+      throw new Error(`Expected NodeAddress, got EdgeAddress: ${x}`);
+    }
+    throw new Error(`Malformed NodeAddress: ${x}`);
+  }
```
Should we check that the input ends with a `SEPARATOR`?

decentralion replied:
Done.

wchargin commented on src/v3/core/graph.test.js:
```diff
@@ -0,0 +1,94 @@
+          });
+        }
+        checkIdentity("on a simple example", ["an", "example"]);
+        describe("with an empty component", () => {
```
Nice, thanks.

decentralion commented:
Thanks for the review comments. I made the following changes, based on your suggestions:
- Switched back to having `X_PREFIX` rather than `BASE_X_ADDRESS`, and now using the cleaner `toXAddress` implementation you suggested
- We now check that valid addresses end with a separator
- Factored out non-exported function `isNodeAddress`, which was convenient once checking involved multiple properties, and will be helpful for the extra assertions in https://github.com/sourcecred/sourcecred/pull/332

wchargin commented on src/v3/core/graph.js:
```diff
@@ -20,22 +21,71 @@ export type NeighborsOptions = {|
+    throw new Error(String(arr));
+  }
+  arr.forEach((s: string) => {
+    if (s.indexOf(SEPARATOR) !== -1) {
```
It would be nice to first check that `s != null`.
